### ISSUE TITLE
Printing changes attributes

### ISSUE THREAD

**Issue by fieker** [oscar-system/Oscar.jl#3653]
```
R = graded_polynomial_ring(QQ, [:x])
quo(R[1], R[2][1]^2)
x = projective_scheme(ans[1])
xx = ProjectiveAlgebraicSet(x);
```
then `X.Xred` is undefined as it should be, but doing
```
julia> xx
```
ie. printing it, will set `X.Xred` after an expensive computation. I suspect this is due to `Oscar.underlying_scheme` being called - which triggers the computation.
The severe bug is that printing should never (substantially) change any structure...
(this confused @wdecker  - as clearly it was set, despite the constructor not doing it)

**Comment by HechtiDerLachs** [comment]
Whether or not `X.Xred` is defined should not be considered to make any difference to the mathematics of `X`. It is mathematically the same object before and after printing, isn't it? So printing does not change `X`. That's like saying `is_prime` changes an ideal, because after calling it, an attribute is set. 

The computation taking long might be an issue of it's own, though. But then printing should probably be rerouted to avoid this computation, if possible.

**Comment by thofma** [comment]
It makes debugging hard, so we usually try to not do it. This is not written down anywhere.

**Comment by HechtiDerLachs** [comment]
I understand. But with all these more high-brow data structures for schemes we rely on more sophisticated caching. And it's really not messing with the internals of `X`, it's just caching an intermediate result. So I think, this is not a bug. 

Unless @fieker still disagrees: Can we close the issue?

**Comment by wdecker** [comment]
```julia
julia> R, (x,) = graded_polynomial_ring(QQ, [:x]);

julia> I = ideal(R, [x^2]);

julia> X = ProjectiveScheme(R, I);

julia> defining_ideal(X)
Ideal generated by
  x^2

julia> XX = ProjectiveAlgebraicSet(X);

julia> defining_ideal(XX)
Ideal generated by
  x

```
@HechtiDerLachs: Can you please explain: Where in the definition of  `ProjectiveAlgebraicSet`does the reduction from x^2 to x happen? And why does it happen  although `is_reduced = = false`? 

```julia
@attributes mutable struct ProjectiveAlgebraicSet{BaseRing<:Field, GradedRingType<:Ring} <: AbsProjectiveAlgebraicSet{BaseRing, GradedRingType}
  X::ProjectiveScheme{BaseRing, GradedRingType}
  Xred::ProjectiveScheme{BaseRing, GradedRingType}
  function ProjectiveAlgebraicSet(X::ProjectiveScheme{S, T}; is_reduced::Bool=false,
                                  check::Bool=true) where {S, T}
    Y = new{S, T}()
    Y.X = X
    if is_reduced
      @check is_geometrically_reduced(X) || error("algebraic sets must be geometrically reduced")
      Y.Xred = X
    else
      # unlock the scheme methods for geometrically reduced schemes.
      set_attribute!(Y, :is_geometrically_reduced, true)
      set_attribute!(Y, :is_reduced, true)
    end
    return Y
  end
end
```

**Comment by HechtiDerLachs** [comment]
@simonbrandhorst worked out the rationale for algebraic sets. If I recall correctly, the idea of an algebraic set is that it should really behave like the geometric set of points. So in particular it should ignore multiplicities and that is what one sees here. 

The `if-else` statement looks a bit bewildering to me. @simonbrandhorst : Could you double check this?

Other than that `defining_ideal` calls `homogeneous_coordinate_ring` which calls `underlying_scheme` which calls `reduced_scheme` which eventually calls `radical` of the original defining ideal for the `fat_scheme`. That's where `x^2` becomes `x` and that's all anticipated by design. 

In general it is rather pointless to track down when and how internal fields are filled. That's not what we had in mind when implementing these things. The user should be provided with convenient functions to ask for whatever they need. Do you have the feeling that easy access to a certain property is missing in this case?

**Comment by HechtiDerLachs** [comment]
It looks to me that the kw-argument `is_reduced` can be set to `true` if one already knows that the input `X` is reduced. Then it doesn't need to compute any radicals later. 

But for the `else`-block: Don't we want to enable these features anyway? And don't we want to do this by implementing `is_reduced(X::AbsProjectiveAlgebraicSet) = true` and similarly for `is_geometrically_reduced`, i.e. make use of the new type instead of attributes? 

Sometimes it is better to do things on both levels (we've just had this with `is_prime`, I think). But I am still a bit confused by this constructor. Even though I think that it is not doing anything fundamentally wrong.

**Comment by fingolfin** [comment]
For the else part I think an accident happened while adjusting the code to use `@check`. Compare:
```julia
@attributes mutable struct ProjectiveAlgebraicSet{BaseRing<:Field, GradedRingType<:Ring} <: AbsProjectiveAlgebraicSet{BaseRing, GradedRingType}
  X::ProjectiveScheme{BaseRing, GradedRingType}
  Xred::ProjectiveScheme{BaseRing, GradedRingType}
  function ProjectiveAlgebraicSet(X::ProjectiveScheme{S, T}; is_reduced::Bool=false,
                                  check::Bool=true) where {S, T}
    Y = new{S, T}()
    Y.X = X
    if is_reduced
      @check is_geometrically_reduced(X) || error("algebraic sets must be geometrically reduced")
      Y.Xred = X
    else
      # unlock the scheme methods for geometrically reduced schemes.
      set_attribute!(Y, :is_geometrically_reduced, true)
      set_attribute!(Y, :is_reduced, true)
    end
    return Y
  end
end
```

with

```julia
@attributes mutable struct AffineAlgebraicSet{BaseRing<:Field, RingType<:MPolyAnyRing} <: AbsAffineAlgebraicSet{BaseRing, RingType}
  X::AffineScheme
  Xred::AffineScheme
  function AffineAlgebraicSet(X::AffineScheme; is_reduced::Bool=false, check::Bool=true)
    A = new{typeof(base_ring(X)), typeof(OO(X))}()
    A.X = X
    if check && is_reduced
      is_geometrically_reduced(X) || error("Algebraic sets must be geometrically reduced")
    end
    if is_reduced
      A.Xred = X
      set_attribute!(A.Xred, :is_geometrically_reduced, true)
      set_attribute!(A.Xred, :is_reduced, true)
      # unlock the scheme methods for geometrically reduced schemes.
    end
    return A
  end
end
```

I think the former code started out looking even more similar to the second function and then was "verschlimmbessert" (no offense intended ;-) ). BTW, `@check` should not be used like that, with `|| error`. I.e.
```
@check is_geometrically_reduced(X) || error("algebraic sets must be geometrically reduced")
```
should be
```
@check is_geometrically_reduced(X) "algebraic sets must be geometrically reduced"
```

**Comment by simonbrandhorst** [comment]
That reduction happens is unintended. The reason is that `underlying_scheme` is called when calling `ambient_space`. Clearly there is no need for any computation to give out the ambient space.  Stevell noticed this correctly and fixed it in his PR. 
As for the `is_reduced` keyword ... that is indeed missleading. A more selfexplaining keyword would be 
`input_is_reduced` I suggest to change the interface accordingly.  
(`input_is_already_reduced` seems like overkill)

**Comment by HechtiDerLachs** [comment]
Resolved by #3657.
